fragment_info.go

```diff
@@ -0,0 +1,418 @@
+	}
+
+	return uint64(cSize), nil
+}
+
+// GetDense checks if a fragment is dense.
+func (fI *FragmentInfo) GetDense(fid uint32) (bool, error) {
+	var cDense C.int32_t
+
+	ret := C.tiledb_fragment_info_get_dense(fI.context.tiledbContext,
+		fI.tiledbFragmentInfo, C.uint32_t(fid), &cDense)
+	if ret != C.TILEDB_OK {
+		return false, fmt.Errorf("Error finding if fragment %d is dense: %s", fid, fI.context.LastError())
+	}
+
+	return cDense == 1, nil
+}
+
+// GetSparse checks if a fragment is sparse.
+func (fI *FragmentInfo) GetSparse(fid uint32) (bool, error) {
+	var cSparse C.int32_t
+
+	ret := C.tiledb_fragment_info_get_sparse(fI.context.tiledbContext,
+		fI.tiledbFragmentInfo, C.uint32_t(fid), &cSparse)
+	if ret != C.TILEDB_OK {
+		return false, fmt.Errorf("Error finding if fragment %d is dense: %s", fid, fI.context.LastError())
+	}
+
+	return cSparse == 1, nil
+}
+
+// GetTimestampRange gets the timestamp range of a fragment.
+func (fI *FragmentInfo) GetTimestampRange(fid uint32) (uint64, uint64, error) {
+	var cStart C.uint64_t
+	var cEnd C.uint64_t
+
+	ret := C.tiledb_fragment_info_get_timestamp_range(fI.context.tiledbContext,
+		fI.tiledbFragmentInfo, C.uint32_t(fid), &cStart, &cEnd)
+	if ret != C.TILEDB_OK {
+		return 0, 0, fmt.Errorf("Error getting the timestamp range for fragment %d: %s", fid, fI.context.LastError())
+	}
+
+	return uint64(cStart), uint64(cEnd), nil
+}
+
+func (fI *FragmentInfo) getNonEmptyDomainSliceFromIndex(did uint32) (
+	*Dimension, interface{}, unsafe.Pointer, error) {
+	config, err := NewConfig()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	context, err := NewContext(config)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	array, err := NewArray(context, fI.uri)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	dimension, tmpDimension, tmpDimensionPtr, err := array.GetNonEmptyDomainSliceFromIndex(uint(did))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return dimension, tmpDimension, tmpDimensionPtr, nil
+}
+
+func (fI *FragmentInfo) getNonEmptyDomainSliceFromName(did string) (
+	*Dimension, interface{}, unsafe.Pointer, error) {
+	config, err := NewConfig()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	context, err := NewContext(config)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	array, err := NewArray(context, fI.uri)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	dimension, tmpDimension, tmpDimensionPtr, err := array.GetNonEmptyDomainSliceFromName(did)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	return dimension, tmpDimension, tmpDimensionPtr, nil
+}
+
+// GetNonEmptyDomainFromIndex retrieves the non-empty domain from a given fragment for a given
+// dimension index.
+// func (fI *FragmentInfo) GetNonEmptyDomainFromIndex
+func (fI *FragmentInfo) GetNonEmptyDomainFromIndex(fid uint32, did uint32) (*NonEmptyDomain, error) {
+	dimension, tmpDimension, tmpDimensionPtr, err := fI.getNonEmptyDomainSliceFromIndex(did)
```

Comment:
@Shelnutt2, do you think that this is a good approach? I tried to reuse `MakeSlice`, but since we do not know the datatype of the dimension, i i ended up refactoring array code (and using `NewArray` eventually)

Reply:
I understand the difficulty here, I don't have a better recommendation. However this approach is not good if the the array is a TileDB-Cloud array. If its a TileDB Cloud array this is going to cause the user to fetch the schema from TileDB Cloud for every fragment it asks for because we end up creating a new Array object each time.

Reply:
Maybe get array only once, caching following calls to array schema in the context of FragmentInfo, as it is changed now?